File: Practica3/Cola/python/Cola.py
class Cola:
    def __init__(self, n):
        self.__arreglo = [0] * (n + 1)
        self.__inicio = 0
        self.__fin = 0
        self.__n = n
    def insert(self, e):
        if self.__fin == self.__n:
            print("Cola Llena")
        else:
            self.__fin = self.__fin + 1
            self.__arreglo[self.__fin] = e
    def remove(self):
        if self.__inicio == 0 and self.__fin == 0:
            print("Cola Vacia")
            return None
        else:
            self.__inicio = self.__inicio + 1
            dato = self.__arreglo[self.__inicio]
            if self.__inicio == self.__fin:
                self.__inicio = 0
                self.__fin = 0
            return dato
    def peek(self):
        if self.isEmpty():
            return None
        return self.__arreglo[self.__inicio + 1]
    def isEmpty(self):
        return self.__inicio == 0 and self.__fin == 0
    def isFull(self):
        return self.__fin == self.__n
    def nroelem(self):  # size
        aux = Cola(self.__n)
        contador = 0
        while not self.isEmpty():
            dato = self.remove()
            aux.insert(dato)
            contador = contador + 1
        while not aux.isEmpty():
            dato = aux.remove()
            self.insert(dato)
        return contador
    def nroelem2(self):  # size
        return self.__fin - self.__inicio
    def __str__(self):
        if self.isEmpty():
            return "Cola Vacía"
        elementos = []
        aux = Cola(self.__n)
        while not self.isEmpty():
            dato = self.remove()
            elementos.append(str(dato))
            aux.insert(dato)
        while not aux.isEmpty():
            dato = aux.remove()
            self.insert(dato)
        return " ".join(elementos)
    def copy(self):
        nueva_cola = Cola(self.__n)
        aux = Cola(self.__n)
        while not self.isEmpty():
            dato = self.remove()
            nueva_cola.insert(dato)
            aux.insert(dato)
        while not aux.isEmpty():
            dato = aux.remove()
            self.insert(dato)
        return nueva_cola
# Ejemplo de uso
colas = Cola(9)
colas.insert(2)
colas.insert(4)
colas.insert(6)
colas.insert(8)
colas.insert(10)
colas.insert(12)
colas.insert(14)
colas.insert(16)
colas.insert(18)
print("Cola:",colas)
colas.insert(20)
aux = colas.copy()
print("Datos de mi Cola:",colas)
print("Cola Auxiliar:", aux)

File: Practica3/Cola/java/src/dates/Cola.java
package dates;

public class Cola {
    private long arreglo[];
    private int inicio;
    private int fin;
    private int n;

    public Cola(int n) {
        this.arreglo = new long[n + 1];
        this.inicio = 0;
        this.fin = 0;
        this.n = n;
    }

    public void insert(long e) {
        if (this.fin == n) {
            System.out.println("Cola Llena");
        } else {
            this.fin = this.fin + 1;
            this.arreglo[this.fin] = e;
        }
    }

    public long remove() {
        if (this.inicio == 0 && this.fin == 0) {
            System.out.println("Cola vacia");
            return -1;
        } else {
            this.inicio = this.inicio + 1;
            long dato = this.arreglo[this.inicio];

            if (this.inicio == this.fin) {
                this.inicio = 0;
                this.fin = 0;
            }
            return dato;
        }
    }

    public long peek() {
        return this.arreglo[this.inicio + 1];
    }

    public boolean isEmpty() {
        return this.inicio == 0 && this.fin == 0;
    }

    public boolean isFull() {
        return this.fin == this.n;
    }

    public int nroelem() { // size
        return this.fin - this.inicio;
    }

    public void mostrarColaHorizontal() {
        for (int i = inicio + 1; i <= fin; i++) {
            System.out.print(arreglo[i] + " ");
        }
        System.out.println();
    }

    public Cola copiarCola() {
        Cola nuevaCola = new Cola(this.n);
        for (int i = inicio + 1; i <= fin; i++) {
            nuevaCola.insert(arreglo[i]);
        }
        return nuevaCola;
    }

    public static void main(String[] args) {
        Cola cola = new Cola(6);
        cola.insert(1);
        cola.insert(3);
        cola.insert(5);
        cola.insert(7);
        cola.insert(9);
        cola.insert(11);

        // Mostrar la cola original de manera horizontal
        System.out.print("Cola : ");
        cola.mostrarColaHorizontal();
        cola.insert(13);
        // Copiar la cola original a una nueva cola
        Cola colaExtra = cola.copiarCola();

        // Mostrar la cola original nuevamente para verificar que no fue afectada
        System.out.print("Datos de mi Cola: ");
        cola.mostrarColaHorizontal();
        System.out.print("Cola Extra: ");
        colaExtra.mostrarColaHorizontal();
    }
}